cmd/backup/script.go

```diff
@@ -318,44 +319,79 @@ func newScript() (*script, error) {
 	return s, nil
 }
 
-// stopContainers stops all Docker containers that are marked as to being
+type noopWriteCloser struct {
+	io.Writer
+}
+
+func (noopWriteCloser) Close() error {
+	return nil
+}
+
+// stopContainersAndServices stops all Docker containers that are marked as to being
 // stopped during the backup and returns a function that can be called to
 // restart everything that has been stopped.
-func (s *script) stopContainers() (func() error, error) {
+func (s *script) stopContainersAndServices() (func() error, error) {
 	if s.cli == nil {
 		return noop, nil
 	}
 
-	allContainers, err := s.cli.ContainerList(context.Background(), types.ContainerListOptions{})
+	dockerInfo, err := s.cli.Info(context.Background())
 	if err != nil {
-		return noop, fmt.Errorf("stopContainers: error querying for containers: %w", err)
+		return noop, fmt.Errorf("(*script).stopContainersAndServices: error getting docker info: %w", err)
 	}
+	isDockerSwarm := dockerInfo.Swarm.LocalNodeState != "inactive"
+	discardWriter := &noopWriteCloser{io.Discard}
 
-	containerLabel := fmt.Sprintf(
+	filterMatchLabel := fmt.Sprintf(
 		"docker-volume-backup.stop-during-backup=%s",
 		s.c.BackupStopContainerLabel,
 	)
+
+	allContainers, err := s.cli.ContainerList(context.Background(), types.ContainerListOptions{})
+	if err != nil {
+		return noop, fmt.Errorf("(*script).stopContainersAndServices: error querying for containers: %w", err)
+	}
 	containersToStop, err := s.cli.ContainerList(context.Background(), types.ContainerListOptions{
```

Comment:
What happens if a user labels a container with `docker-volume-backup.stop-during-backup` which is part of a service and also adds the same label to the service itself?

Reply:
I was wondering the same, yes. Documentation says do one or the other, so one option would be to say "tough luck, please do what is described". Another option would be checking if a container has a parent service, and in case it does check if it is labeled as well. In case there is a conflict, bail.

Probably worth doing this extra check.

Reply:
What I think would happen in practice is:
- the container is stopped
- its parent service is scaled down, removing the container
- the script tries to restart a now-removed container and fails

Reply:
If possible I would aim to be user friendly, while it would be hard to accidentally configure it this way, still a possibility.

Instead of checking if the container has a parent service with the labels, the other way might be easier to implement. ( Check if a labelled service has any also labelled containers. )